gestalt-asset-core/src/main/java/org/terasology/gestalt/assets/AssetType.java

```diff
@@ -288,15 +269,45 @@ synchronized void registerAsset(Asset<U> asset, DisposalHook disposer) {
         if (closed) {
             throw new IllegalStateException("Cannot create asset for disposed asset type: " + assetClass);
         } else {
-            if (asset.getUrn().isInstance()) {
-                instanceAssets.put(asset.getUrn(), new WeakReference<>(assetClass.cast(asset)));
-            } else {
-                loadedAssets.put(asset.getUrn(), assetClass.cast(asset));
+            if(!asset.getUrn().isInstance()) {
+                loadedAssets.put(asset.getUrn(), new SoftReference<T>(assetClass.cast(asset)));
             }
             references.add(new AssetReference<>(asset, disposalQueue, disposer));
         }
     }
 
+    /**
+     * Notifies the asset type when an asset is disposed
+     *
+     * @param asset The asset that was disposed.
+     */
+    void onAssetDisposed(Asset<U> asset) {
+        if (!asset.getUrn().isInstance()) {
+            compactResource(asset.getUrn());
+        }
+    }
+
+    private void compactResource(ResourceUrn target) {
+        Preconditions.checkArgument(!target.isInstance());
+        if(!loadedAssets.containsKey(target)) {
+            return;
+        }
+
+        Reference<T> reference = loadedAssets.get(target);
+        Asset<U> current = reference.get();
+        if(current != null && current.isDisposed() && Iterables.isEmpty(current.instances())) {
+            logger.warn("non instanced asset is disposed with instances. instances will become orphaned.");
+        }
+
+        if(current == null || current.isDisposed()) {
+            // disposing of a non instanced asset will orphan the instanced assets.
+            loadedAssets.remove(target);
+            return;
+        }
+        current.cleanup();
```

Comment:
This line probably is unreachable without an asset being loaded fresh into the urn in the meantime. Which I guess is possible through the onAssetDisposed pathway, since it isn't synchronized. Not sure it is a meaningful action regardless.